database.go

```diff
@@ -0,0 +1,723 @@
+// The two collections must already be sorted.
+func (l SeriesIDs) Reject(r SeriesIDs) SeriesIDs {
+	var i, j int
+
+	ids := make([]uint32, 0, len(l))
+	for i < len(l) && j < len(r) {
+		if l[i] == r[j] {
+			i += 1
+			j += 1
+		} else if l[i] < r[j] {
+			ids = append(ids, l[i])
+			i += 1
+		} else {
+			j += 1
+		}
+	}
+
+	// append the remainder
+	if i < len(l) {
+		ids = append(ids, l[i:]...)
+	}
+
+	return SeriesIDs(ids)
+}
+
+// AddSeries adds the series for the given measurement to the index. Returns false if already present
+func (t *database) AddSeries(name string, s *Series) bool {
+	// if there is a measurement for this id, it's already been added
+	if t.seriesToMeasurement[s.ID] != nil {
+		return false
+	}
+
+	// get or create the measurement index and index it globally and in the measurement
+	idx := t.createMeasurementIfNotExists(name)
+
+	t.seriesToMeasurement[s.ID] = idx
+	t.series[s.ID] = s
+
+	// TODO: add this series to the global tag index
+
+	return idx.addSeries(s)
+}
+
+// createMeasurementIfNotExists will either add a measurement object to the index or return the existing one.
+func (t *database) createMeasurementIfNotExists(name string) *Measurement {
+	idx := t.measurements[name]
+	if idx == nil {
+		idx = NewMeasurement(name)
+		t.measurements[name] = idx
+		t.names = append(t.names, name)
+		sort.Strings(t.names)
+	}
+	return idx
+}
+
+// AddField adds a field to the measurement name. Returns false if already present
+func (t *database) AddField(name string, f *Field) bool {
+	panic("not implemented")
+	return false
+}
+
+// MeasurementsForSeriesIDs returns a collection of unique Measurements for the passed in SeriesIDs.
+func (t *database) MeasurementsForSeriesIDs(seriesIDs SeriesIDs) []*Measurement {
+	measurements := make(map[*Measurement]bool)
+
+	for _, id := range seriesIDs {
+		measurements[t.seriesToMeasurement[id]] = true
+	}
+
+	values := make([]*Measurement, 0, len(measurements))
+	for m, _ := range measurements {
+		values = append(values, m)
+	}
+
+	return values
+}
+
+// SeriesIDs returns an array of series ids for the given measurements and filters to be applied to all.
+// Filters are equivalent to and AND operation. If you want to do an OR, get the series IDs for one set,
+// then get the series IDs for another set and use the SeriesIDs.Union to combine the two.
+func (t *database) SeriesIDs(names []string, filters Filters) SeriesIDs {
+	// they want all ids if no filters are specified
+	if len(filters) == 0 {
+		ids := SeriesIDs(make([]uint32, 0))
+		for _, idx := range t.measurements {
+			ids = ids.Union(idx.ids)
+		}
+		return ids
+	}
+
+	ids := SeriesIDs(make([]uint32, 0))
+	for _, n := range names {
+		ids = ids.Union(t.seriesIDsForName(n, filters))
+	}
+
+	return ids
+}
+
+// TagKeys returns a sorted array of unique tag keys for the given measurements.
+func (t *database) TagKeys(names []string) []string {
+	if len(names) == 0 {
```

Comment:
OK, got it.

I think the comment could be improved in that case. E.g.
```
TagKeys returns all keys in the database. If the names slice is non-empty however, then only keys for those Measurement names are returned.
```
Just a suggestion.